domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -161,101 +192,305 @@ class TopicListController @Inject constructor(
+    return recentlyPlayedStoryList.size + olderPlayedStoryList.size + suggestedStoryList.size
+  }
+
+  private fun PromotedStoryList.Builder.addAllUpTo(
+    iterable: Iterable<PromotedStory>,
+    addAll: PromotedStoryList.Builder.(Iterable<PromotedStory>) -> PromotedStoryList.Builder,
+    limit: Int
+  ): PromotedStoryList.Builder {
+    return this.addAll(iterable.take(limit - this.getTotalPromotedStoryCount()))
+  }
+
+  private fun populateRecentlyPlayedStories(
+    topicProgressList: List<TopicProgress>
+  ): List<PromotedStory> {
+    computePromotedStories(topicProgressList).let {
+      Pair(it.first, it.second)
+      if (it.second < ONE_WEEK_IN_DAYS)
+        return it.first
+    }
+    return emptyList()
+  }
+
+  private fun populateOlderPlayedStories(
     topicProgressList: List<TopicProgress>
-  ): OngoingStoryList {
-    val ongoingStoryListBuilder = OngoingStoryList.newBuilder()
-    topicProgressList.forEach { topicProgress ->
+  ): List<PromotedStory> {
+    computePromotedStories(topicProgressList).let {
+      Pair(it.first, it.second)
+      if (it.second > ONE_WEEK_IN_DAYS)
+        return it.first
+    }
+    return emptyList()
+  }
+
+  private fun computePromotedStories(
+    topicProgressList: List<TopicProgress>
+  ): Pair<List<PromotedStory>, Long> {
+    var numberOfDaysPassed = 0L
+    val recentlyPlayedPromotedStoryList = mutableListOf<PromotedStory>()
+    val sortedTopicProgressList =
+      topicProgressList.sortedByDescending { it.lastPlayedTimestamp }
+
+    sortedTopicProgressList.forEach { topicProgress ->
       val topic = topicController.retrieveTopic(topicProgress.topicId)
+
       topicProgress.storyProgressMap.values.forEach { storyProgress ->
         val storyId = storyProgress.storyId
         val story = topicController.retrieveStory(topic.topicId, storyId)
 
-        val completedChapterProgressList =
-          storyProgress.chapterProgressMap.values
-            .filter { chapterProgress ->
-              chapterProgress.chapterPlayState ==
-                ChapterPlayState.COMPLETED
-            }
-            .sortedByDescending { chapterProgress -> chapterProgress.lastPlayedTimestamp }
-
-        val lastCompletedChapterProgress: ChapterProgress? =
+        val completedChapterProgressList = getCompletedChapterProgressList(storyProgress)
+        val mostRecentCompletedChapterProgress: ChapterProgress? =
           completedChapterProgressList.firstOrNull()
 
-        val startedChapterProgressList =
-          storyProgress.chapterProgressMap.values
-            .filter { chapterProgress ->
-              chapterProgress.chapterPlayState ==
-                ChapterPlayState.STARTED_NOT_COMPLETED
-            }
-            .sortedByDescending { chapterProgress -> chapterProgress.lastPlayedTimestamp }
-
+        val startedChapterProgressList = getStartedChapterProgressList(storyProgress)
         val recentlyPlayerChapterProgress: ChapterProgress? =
```

Comment:
This should be named similarly to the other variable. Perhaps: ``mostRecentStartedChapterProgress`` would make things much clearer (or 'least' if the one above isn't named correctly).

Reply:
`mostRecentStartedChapterProgress` won't sound good as we can have recent or old chapters in progress. Also I have changed the function name to `computePlayedStories`. Based on this if we can name it something else which will give it's clear context.

Reply:
Could we use 'latestStartedChapterProgress' / 'oldestStartedChapterProgress' instead? I'm actually still confused which it is so I think the clarification in naming is important here.

Reply:
Yeah right! `latestStartedChapterProgress` sounds good to me.

Reply:
> Yeah right! `latestStartedChapterProgress` and `latestCompletedChapterProgress` sounds fine or just `startedChapterProgress` or `completedChapterProgress`, because at the end based on its timestamp we are deciding if it was recent or oldest.